src/io/binfile_writer.cc

```diff
@@ -0,0 +1,142 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "singa/io/binfile_writer.h"
+#include "singa/utils/logging.h"
+
+namespace singa {
+namespace io {
+bool BinFileWriter::Open(const std::string& path, Mode mode, int capacity) {
+  CHECK(!fdat_.is_open());
+  path_ = path;
+  mode_ = mode;
+  capacity_ = capacity;
+  buf_ = new char[capacity_];
+  switch (mode) {
+    case kCreate:
+      fdat_.open(path_, std::ios::binary | std::ios::out | std::ios::trunc);
+      CHECK(fdat_.is_open()) << "Cannot create file " << path_;
+      break;
+    case kAppend:
+      fdat_.open(path_, std::ios::in | std::ios::binary);
+      if (!fdat_.is_open()) {
+        // path may be a directory
+        path_ = path + "/shard.dat";
+        fdat_.open(path_, std::ios::in | std::ios::binary);
+      }
+      CHECK(fdat_.is_open()) << "Cannot open file " << path_;
+      fdat_.close();
+      {
+        int last_tuple = PrepareForAppend(path_);
+        fdat_.open(path_, std::ios::binary | std::ios::out | std::ios::in |
+                              std::ios::ate);
+        fdat_.seekp(last_tuple);
+      }
+      break;
+    default:
+      LOG(FATAL) << "unknown model to open KVFile " << mode;
+      break;
+  }
+  return fdat_.is_open();
+}
+
+void BinFileWriter::Close() {
+  if (mode_ == kCreate || mode_ == kAppend) Flush();
+  if (buf_ != nullptr) {
+    delete buf_;
+    buf_ = nullptr;
+  }
+  if (fdat_.is_open()) fdat_.close();
+}
+
+bool BinFileWriter::Write(const std::string& key, const std::string& value) {
+  CHECK(fdat_.is_open()) << "File not open!";
+  if (value.size() == 0) return false;
+  // magic_word + (key_len + key) + val_len + val
+  char magic[4];
+  int size;
+  magic[0] = 's';
```

Comment:
define the MAGIC_WORD in the .h file as a constant, e.g.,

const char kMagicWord[]= "sg“;